src/r3/atomic-swap/evm/src/SwapVault.sol

```diff
@@ -0,0 +1,185 @@
+ ******************************************************************************/
+
+pragma solidity 0.8.20;
+
+import "openzeppelin/token/ERC20/IERC20.sol";
+import "openzeppelin/token/ERC20/extensions/IERC20Metadata.sol";
+import "openzeppelin/token/ERC20/utils/SafeERC20.sol";
+import "openzeppelin/token/ERC721/IERC721.sol";
+import "openzeppelin/token/ERC1155/IERC1155.sol";
+import "openzeppelin/token/ERC721/utils/ERC721Holder.sol";
+import "openzeppelin/token/ERC1155/utils/ERC1155Holder.sol";
+
+library Utils {
+    function safeSupportsInterface(address addr, bytes4 interfaceId) external view returns (bool supported) {
+        (bool success, bytes memory result) = addr.staticcall(
+                abi.encodeWithSelector(IERC165.supportsInterface.selector, interfaceId)
+            );
+        
+        supported = success && result.length > 0 && abi.decode(result, (bool));
+    }
+}
+
+contract SwapVault is ERC721Holder, ERC1155Holder {
+
+    error ZERO_ADDRESS();
+    error COMMIT_TO_SELF();
+    error ALREADY_COMMITTED();
+    error INVALID_AMOUNT();
+    error INVALID_STATUS();
+    error INVALID_CLAIMER();
+    error UNSUPPORTED_INTERFACE();
+
+    event Commit(string indexed swapId, bytes32 holdHash);
+    event Transfer(string indexed swapId, bytes32 holdHash);
+    event Revert(string indexed swapId, bytes32 holdHash);
+
+    using Utils for address;
+
+    struct Commitment {
+        string swapId;
+        uint256 status;
+        address owner;
+        address recipient;
+        uint256 amount;
+        uint256 tokenId;
+        address tokenAddress;
+        uint256 signaturesThreshold;                
+        bytes32[] signatures;
+    }
+
+    mapping(string => Commitment) _committmentState;
+
+    function commit(string calldata swapId, address recipient, uint256 signaturesThreshold) external {
+        _commit(swapId, recipient, signaturesThreshold);
+    }
+
+    function commitWithToken(string calldata swapId, address tokenAddress, uint256 amount, address recipient, uint256 signaturesThreshold) external {
+        Commitment storage commitment =  _commit(swapId, recipient, signaturesThreshold);
+
+        commitment.amount = amount;
+        commitment.tokenAddress = tokenAddress;
+
+        SafeERC20.safeTransferFrom(IERC20(commitment.tokenAddress), msg.sender, address(this), commitment.amount);
+
+        emit Commit(swapId, commitmentHash(swapId));
+    }
+
+    function commitWithToken(string calldata swapId, address tokenAddress, uint256 tokenId, uint256 amount, address recipient, uint256 signaturesThreshold) external {
+        Commitment storage commitment = _commit(swapId, recipient, signaturesThreshold);
+
+        commitment.amount = amount;
+        commitment.tokenId = tokenId;
+        commitment.tokenAddress = tokenAddress;
+
+        if(IERC165(tokenAddress).supportsInterface(0x80ac58cd)) { // ERC721
+            // 
+            if(commitment.amount != 1) revert INVALID_AMOUNT();
```

Comment:
Could be wrong but this seems like a more appropriate location for a assert then a revert 

https://www.linkedin.com/pulse/require-vs-revert-assert-solidity-arslan-maqbool/

Reply:
@ILoomans require and revert both result in transaction being reverted. Gas spent to "get there" is lost.

I honestly never saw a project (at least looking at well know and not so well known protocols) that uses assert but besides that I use if cond / revert / error as it returns required and clear information about the error when the transaction fail and it uses the least gas due to the fact that errors are translated into 4 bytes hash instead of requiring a whole string to be both stored with the contract and allocated on the stack.